### PR TITLE
Spatially partition frozen actors

### LINKED ISSUE #9614: Use spatial partitioning to speed up frozen actors
One of our biggest perf hogs is FrozenActorLayer, which checks every frozen actor for visibility every tick.  We could save a lot of perf by tracking dirty flags for each bin in a spatial partition, and only ticking the frozen actors that are in dirty bins.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/27333596-use-spatial-partitioning-to-speed-up-frozen-actors?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by RoosterDragon**
Like other parts of the engine, this PR now tracks change in the shroud so it only had to update frozen actors on the map in areas where the shroud may have changed, rather than needing to update them all.

On the RA shellmap, this reduces CPU time spent in FrozenActorLayer.Tick from 2.32% to 1.51%. (For comparison - this is our most expensive trait to tick at the moment, the next worst is RenderSprites at 1.46%)

Closes #9614.

### CODE REVIEW

File: OpenRA.Game/Traits/Player/FrozenActorLayer.cs [hunk 1]
```diff
@@ -146,54 +148,123 @@ public class FrozenActorLayer : IRender, ITick, ISync
...
 
 		public void Add(FrozenActor fa)
 		{
-			frozen.Add(fa.ID, fa);
+			frozenActorsById.Add(fa.ID, fa);
 			world.ScreenMap.Add(owner, fa);
+			partitionedFrozenActorIds.Add(fa.ID, FootprintBounds(fa));
+		}
+
+		Rectangle FootprintBounds(FrozenActor fa)
```
Comment (RoosterDragon) [inline OpenRA.Game/Traits/Player/FrozenActorLayer.cs]: @pchote: This appears to work in TS - but can you confirm if this is a valid way to get the bounds in isometric space?

Comment (pchote) [inline OpenRA.Game/Traits/Player/FrozenActorLayer.cs]: Haven't tested it, but the underlying logic at least is sound.

**Comment by Mailaender** [comment]
Can't spot anything wrong in-game. :white_check_mark:

File: OpenRA.Game/Traits/Player/FrozenActorLayer.cs [hunk 2]
```diff
@@ -146,54 +148,123 @@ public class FrozenActorLayer : IRender, ITick, ISync
...
+			partitionedFrozenActorIds = new SpatiallyPartitioned<uint>(
+				world.Map.MapSize.X, world.Map.MapSize.Y, binSize);
+			dirtyBins = new bool[world.Map.MapSize.X / binSize + 1, world.Map.MapSize.Y / binSize + 1];
+			self.Trait<Shroud>().CellsChanged += cells =>
+			{
+				foreach (var cell in cells)
+				{
+					var x = cell.U / binSize;
+					var y = cell.V / binSize;
+					dirtyBins[x, y] = true;
```
Comment (pchote) [inline OpenRA.Game/Traits/Player/FrozenActorLayer.cs]: You can save a bit more perf (unnecessary bounds checking) by making this a 1d array indexed on `y * world.Map.MapSize.X + x`.

Comment (RoosterDragon) [inline OpenRA.Game/Traits/Player/FrozenActorLayer.cs]: Added, but with the correct dimensions.

**Comment by abcdefg30** [comment]
Looks good to me, too. :+1: / :white_check_mark:

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/61ee467173d19078eb221d21c9a15ba167f9ad89...1eee8270bd3d9989efa85c9d42883bddc056ad3d)

**Pull request merged**